From be6bd904b77c850819a3eddd041efeb1b59feccf Mon Sep 17 00:00:00 2001
From: Jan Beich <jbeich@FreeBSD.org>
Date: Sat, 11 Jan 2020 21:18:59 +0000
Subject: [PATCH] Get thread ID on more platforms

media_driver/agnostic/common/cm/cm_task_internal.cpp:896:39: error: use of undeclared identifier 'SYS_gettid'
    m_taskProfilingInfo.threadID    = CmGetCurThreadId(); // Get Thread ID
                                      ^
media_driver/linux/common/cm/hal/cm_innerdef_os.h:98:37: note: expanded from macro 'CmGetCurThreadId'
 #define CmGetCurThreadId()  syscall(SYS_gettid)
                                     ^
media_driver/agnostic/common/cm/cm_task_internal.cpp:997:39: error: use of undeclared identifier 'SYS_gettid'
    m_taskProfilingInfo.threadID    = CmGetCurThreadId(); // Get Thread ID
                                      ^
media_driver/linux/common/cm/hal/cm_innerdef_os.h:98:37: note: expanded from macro 'CmGetCurThreadId'
 #define CmGetCurThreadId()  syscall(SYS_gettid)
                                     ^
---
 .../linux/common/cm/hal/cm_innerdef_os.h      | 21 +++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/media_driver/linux/common/cm/hal/cm_innerdef_os.h b/media_driver/linux/common/cm/hal/cm_innerdef_os.h
index ecc5f7aef1..03a5524696 100644
--- a/media_driver/linux/common/cm/hal/cm_innerdef_os.h
+++ b/media_driver/linux/common/cm/hal/cm_innerdef_os.h
@@ -36,7 +36,15 @@
 #include "mos_os.h"
 #include "media_libva_common.h"
 #include <sys/types.h>
+#if defined(__linux__)
 #include <sys/syscall.h>
+#elif defined(__DragonFly__) || defined(__FreeBSD__)
+#include <pthread_np.h>
+#elif defined(__NetBSD__)
+#include <lwp.h>
+#elif defined(__sun)
+#include <thread.h>
+#endif
 #include <unistd.h>
 
 
@@ -86,5 +94,18 @@ inline void GetLocalTime(PSYSTEMTIME psystime)
 #endif
 
 #define CmGetCurProcessId() getpid()
+#if defined(__linux__)
 #define CmGetCurThreadId()  syscall(SYS_gettid)
+#elif defined(__DragonFly__) || defined(__FreeBSD__)
+#define CmGetCurThreadId()  pthread_getthreadid_np()
+#elif defined(__NetBSD__)
+#define CmGetCurThreadId()  _lwp_self()
+#elif defined(__OpenBSD__)
+#define CmGetCurThreadId()  getthrid()
+#elif defined(__sun)
+#define CmGetCurThreadId()  thr_self()
+#else
+#warning "Cannot get kernel thread identifier on this platform."
+#define CmGetCurThreadId()  (uintptr_t)pthread_self()
+#endif
 
